### PR TITLE
Add arcsde-sql directory and a .sql file

### PR DESCRIPTION
**Pull request by natas333x2**
Add the arcsde-sql directory for future SQL.  Inside the arcsde-sql dir is an SQL script to create postgres geometry points from an external table containing an X and Y column.

### CODE REVIEW

**Comment by jgravois** [comment]
thanks @natas333x2 

in the interest of organization, we like to place individual code samples within their own folder and include a README.md alongside that explicitly defines the scope of the sample and the general steps to run it.

you guys would know better than me, but because of differences in the flavor of sql used in Oracle, SQL Server and PostGres databases, it would probably be good to create subfolders inside a parent `sql` folder to organize content that way too.

does all that sound good?  you can continue working in the same branch and just lay down additional commits.  afterward, we need to get someone who isn't skurred to guinea pig running your code prior to merging.

**Comment by jgravois** [comment]
your atest changes look good!  now i'm just being nitpicky, but since we're setting a precedent here, what the heck.

i'd say that the [python](https://github.com/Esri/developer-support/tree/master/arcpy-python) folder makes the best corollary with SQL scripts in that there is typically only going to be a single sample file and that its safe to trim its name significantly in comparison to what is used to name the folder.

i thought initially that it would be most helpful to have a single `postgres` folder to contain any sample code aimed at that database (and skip including postgres again in either the individual sample folder or script name), but i'm fine with skipping individual folders for different databases and just prefacing the folder name with a standardized convention for the database like you are already doing too.

basically all that is just a really long way of saying, 'this works for me, just get rid of `postgres` in the script filename.

the only other thing left to do prior to merging is to have someone run the code and confirm that it has the desired effect and that the instructions provided give sufficient explanation.  if an SDE 'expert' is available, that would be *okay*, but i think as a general rule it is preferable when the person doing the QA/code review is less comfortable w/ the technology because this gives us better insight into whether the sample is actually self explanatory.

**Comment by natas333x2** [comment]
Thanks @jgravois.  Maybe we can brainstorm how to house the different DBMS flavors.  I like your idea of separate folders for each vendor then omitting the name from the file name.  Or even a blanket sql folder and each folder should specify the vendor in the name?  I'm open to any format.

Also, what are your thoughts on adding sample data?  For example, the postgresql-xy-cursor.sql file requires a table full of XY values.  I made that known in the readme but unless someone has that kind of data, in postgres, testing might be difficult.

**Comment by jgravois** [comment]
i vote:
* blanket sql folder
* folder for each RDMS
* terse filenames

please refactor in a new commit and lay it down directly on top of what you've already done.  hopefully the new structure will encourage folks to get involved with translating existing samples into other flavors when necessary.

in the interest of minimizing bloat, we generally try to avoid including sample data directly in the repository whenever possible.  as an alternative, you could either document the expected syntax in the readme, link to external files hosted on arcgis.com.  when the tester attempts to run your sample code, this gives us a chance to see whether the info you have provided about the data itself is sufficient or not.

**Comment by jgravois** [comment]
@natas333x2 let me know if you need any help implementing the changes i suggested last week.

**Comment by jgravois** [comment]
sorry for the delay in getting back to this, but since you didn't write any additional messages in this Pull Request, i didn't get a notification that you'd laid down additional commits.

looks great!

a couple things though:
1. your sample for creating the new table and inserting rows doesn't work *as is* because you aren't declaring an objectid field..

```
CREATE TABLE sde.test_xy
(
name character varying(255),
x_coord numeric(38,8),
y_coord numeric(38,8)
)
...

INSERT INTO sde.test_xy(name, x_coord, y_coord) 
VALUES ( 'Iowa Heartland Development', 1601700.38827856, 592924.23589906, 1);
```
the `1` throws an error (i'm assuming it should be a field called 'objectid' above, not sure about the datatype.

i tried running your actual script in a plain POSTGIS enabled PostGres database (no SDE) substituting `public` as the schema (and ensuring that i stuck with `pg_geometry`), but that script through an error too because i didn't already have a `public.points_pggeom` and `public.next_rowid` table defined.  is SDE definitely mandatory or am i just missing some more setup steps?

**Comment by natas333x2** [comment]
Thanks for catching the ID value.  I removed the values from the insert statements as they play no role in the cursor statement.  

Regarding the sde.next_rowid function, this definitely requires an enterprise geodatabase.  The functions would be owned by the sde schema.  The points_pggeom table is the required feature class that needs to exist before the cursor is run.  This is written in the readme file but may not be clear when you look at the cursor.  I'll add some comments to the cursor to explain that this is the target.

Oh, and sorry about all the commits.  I'll learn to squash them before I add anything new. ;)

**Comment by jgravois** [comment]
no need to squash until the very end.  I'll take another crack at it tomorrow and see if I have anymore feedback.  either way, really stoked to get some SQL in here!

**Comment by jgravois** [comment]
today i laid down a new 9.3 Postgres instance and ran the `Create Enterprise GDB` GP tool.  afterwards i created a new empty point feature class which can be accessed in ArcMap via
`sdetest.sde.points_pggeom`.

afterward, i ran the updated script to create the new table and insert records and afterward confirmed in PgAdmin that `test_xy` appears alongside the other `sde` schema's tables.

but when i try to run the actual SQL script to inject records into the feature class i encounter the following error:
```
ERROR:  function st_geomfromtext(text, integer) does not exist
LINE 3:       ST_GeomFROMText('POINT(' || cast(x_var as text)|| ' ' ...

HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
```
i never saw an opportunity when i created the feature class to define whether i wanted it to use pg_geometry or ST_Geometry, so i tried running the second script, in hopes that i was obligated to supply the second
```
ERROR:  relation "sde.points_stgeom" does not exist
LINE 1: INSERT INTO sde.points_stgeom (objectid, geom) 
```
even though i can see pretty clearly in PgAdmin that the table is there
screenshot 2015-05-06 17 17 51

**Comment by natas333x2** [comment]
Hi John,

Thanks for hanging in there.  When a Postgres geodatabase is created using the Create Enterprise Geodatabase gp tool, [the PostGIS template is not used](http://resources.arcgis.com/en/help/main/10.2/index.html#//002p0000006v000000#GUID-8208C1FB-C1C4-46C4-BEAA-257796A12661).  To enable a geodatabase to use the PostGIS geometry type, create a new database with PGAdmin using the PostGIS template database.  From there, in ArcGIS, use the Enable Geodatabase gp tool to add the geodatabase.  Many of out PG users don't do this as they intend on using St_Geometry only.  This is the reason for including code for both types.

However, you did point out something I overlooked with the ST_Geometry cursor.  In an effort to keep the types of geometry separate, I wrote the target table as "sde.points_stgeom" instead of sde.points_pggeom".  The comments at the top do state that "both examples" use a target feature class named "sde.points_pggeom".  I corrected this in the script.

For your tests, since you are pretty much ready to go, change ST_Geometry cursor target name to sde.points_stgeom.

**Comment by jgravois** [comment]
i *knew* i was getting close, but looking at it again i should have been able to figure it out myself.  thanks for that.  i also knew i make the perfect guinea pig because i have no idea what im doing. :smile:

everything ran fine after i followed your instructions **and** updated the fieldname to **shape** instead of *geom*

```PLpgSQL
INSERT INTO sde.points_stgeom (objectid, shape)
```
more feedback
1. youre specifying WGS84 as the coordinate system in your script, but your sample XYs aren't lat/long.
2. in terms of usability, we should just add your two actual example SQL scripts inline in the readme.  all .sql thats inline can be attached as supplemental .sql files in the folder as well.
3.  we should specify `PLpgSQL` to get syntax highlighting for the inline codeblocks.  see [this](https://help.github.com/articles/github-flavored-markdown/#syntax-highlighting) article for more info.

**Comment by jgravois** [comment]
laid down my own suggestions and PR'ed against @natas333x2 's fork [here](https://github.com/natas333x2/developer-support/pull/2)

**Comment by natas333x2** [comment]
Nice!  Thanks for helping me out with it.  I learned a lot.  I think what happened is I get stuck in database/SQL mode and think that folks can read my mind.  I'll take my time with the next one.
